binderhub/repoproviders.py

```diff
@@ -186,16 +189,36 @@ class GitRepoProvider(RepoProvider):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        url, resolved_ref = self.spec.rsplit('/', 1)
+        url, unresolved_ref = self.spec.split('/', 1)
         self.repo = urllib.parse.unquote(url)
-        if not resolved_ref:
-            raise ValueError("`resolved_ref` must be specified as a query parameter for the basic git provider")
-        self.sha1_validate(resolved_ref)
-        self.resolved_ref = resolved_ref
-        self.unresolved_ref = resolved_ref
+        self.unresolved_ref = urllib.parse.unquote(unresolved_ref)
+        if not self.unresolved_ref:
+            raise ValueError("`unresolved_ref` must be specified as a query parameter for the basic git provider")
 
     @gen.coroutine
     def get_resolved_ref(self):
+        if hasattr(self, 'resolved_ref'):
+            return self.resolved_ref
+
+        try:
+            # Check if the reference is a valid SHA hash
+            self.sha1_validate(self.unresolved_ref)
+        except ValueError:
+            # The ref is a head/tag and we resolve it using `git ls-remote`
+            command = ["git", "ls-remote", self.repo, self.unresolved_ref]
+            result = subprocess.run(command, text=True, stdout=subprocess.PIPE)
+            if result.returncode:
+                raise RuntimeError("Unable to run git ls-remote to get the `resolved_ref`")
+            if not result.stdout:
+                raise ValueError("The specified branch, tag or commit SHA ('{}') was not found on the remote repository."
+                                .format(self.unresolved_ref))
+            resolved_ref = result.stdout.split(None, 1)[0]
+            self.sha1_validate(resolved_ref)
```

Comment:
Is validating the SHA that is returned by `git ls-remote` too much? I think this safe-guard is nice to have for unexpected output of `git ls-remote`.